platform-sdk/swirlds-common/src/main/java/com/swirlds/common/merkle/proof/algorithms/StateProofTreeBuilder.java

```diff
@@ -0,0 +1,305 @@
+     * the payloads. Nodes are returned in reverse post-ordered depth first traversal order.
+     *
+     * @param merkleRoot the merkle root
+     * @param payloads   the payloads
+     * @return a list of merkle nodes that are ancestors of at least one of the payloads
+     * @throws IllegalStateException if the payloads and the merkle tree are inconsistent with each other
+     */
+    private static List<MerkleNode> getMerkleNodesForStateProofTree(
+            @NonNull final MerkleNode merkleRoot, @NonNull final List<MerkleLeaf> payloads) {
+
+        // Note: we use a reverse post-ordered depth first traversal here intentionally. This causes a node's children
+        // to be popped off the top of a stack in left-to-right order (as opposed to a standard DFS traversal,
+        // which will pop them off in right-to-left order).
+
+        final List<MerkleNode> nodes = new ArrayList<>();
+        merkleRoot
+                .treeIterator()
+                .setOrder(MerkleIterationOrder.REVERSE_POST_ORDERED_DEPTH_FIRST)
+                .ignoreNull(true)
+                .setFilter(node -> isAncestorOfPayload(node, payloads))
+                .setDescendantFilter(node -> isAncestorOfPayload(node, payloads))
+                .forEachRemaining(nodes::add);
+
+        return nodes;
+    }
+
+    /**
+     * Construct a set of merkle routes from a list of merkle nodes.
+     *
+     * @param nodes the nodes to extract the routes from
+     * @return a set of all merkle routes in the list of nodes
+     */
+    private static Set<MerkleRoute> getMerkleRouteSet(@NonNull final List<MerkleNode> nodes) {
+        final Set<MerkleRoute> routes = new HashSet<>();
+        for (final MerkleNode node : nodes) {
+            routes.add(node.getRoute());
+        }
+        return routes;
+    }
+
+    /**
+     * Sanity check the payloads against the node list.
+     *
+     * @param nodes      the list of nodes to include in the state proof tree
+     * @param nodeRoutes the set of routes of the nodes in the node list
+     * @param payloads   the payloads to include in the state proof
+     */
+    private static void validatePayloads(
+            @NonNull final List<MerkleNode> nodes,
+            @NonNull final Set<MerkleRoute> nodeRoutes,
+            @NonNull final List<MerkleLeaf> payloads) {
+
+        // Every payload should appear in the node list.
+        for (final MerkleLeaf leaf : payloads) {
+            if (!nodeRoutes.contains(leaf.getRoute())) {
+                throw new IllegalStateException("Payloads are inconsistent with merkle tree. Payloads contain "
+                        + "leaf at " + leaf.getRoute() + " which is not in the merkle tree.");
+            }
+        }
+
+        // The number of payloads should match exactly the number of leaves in the node list.
+        int leafCount = 0;
+        for (final MerkleNode node : nodes) {
```

Comment:
Any reason why not use `Stream.count()`?

Reply:
No particular reason other than personal preference. For really simple things like this, a for loop seems like a clean way to do it (to me at least).